### ISSUE TITLE
PathConvertor doesn't match new line characters

### ISSUE THREAD

**Issue by and-semakin** [Kludex/starlette#1058]
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`. (Checked on `starlette==0.13.8`).
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

Let's imagine that I pass some text data as a path parameter and this data can generally contain any Unicode character. It turned out that it's impossible to pass the data with newline characters (`\n`) inside. Starlette decodes the URL somewhere before it tries to match routes, so it doesn't help if you try to pass this character in encoded form like `%0A`. It either doesn't match the route or cuts the new line character off (but it breaks the application logic). Seems like the issue is in [the regex](https://github.com/encode/starlette/blob/41218ac546a3c7a0f1982f2677177ce9a0ab3616/starlette/convertors.py#L30) used by the `PathConvertor`:

```python
class PathConvertor(Convertor):
    regex = ".*"
```

According [to many sources](https://riptutorial.com/regex/example/32238/why-doesn-t-dot-----match-the-newline-character----n---) on the internet, `.` (the dot character) matches any character except for `\n`. Should we allow it to match literally any character?

Possible solutions:

* replace `.` with `\s\S` — it's a popular trick to match ANY character including new lines:

```python
class PathConvertor(Convertor):
    regex = "[\s\S]*"
```

* put the `re.DOTALL` flag somewhere, for example [here](https://github.com/encode/starlette/blob/41218ac546a3c7a0f1982f2677177ce9a0ab3616/starlette/routing.py#L123):

```python
return re.compile(path_regex, flags=re.DOTALL), path_format, param_convertors
```

Notice that in both cases newline characters will appear in the matched data (it changes the current behavior but I'm not sure how big the impact will be).

I understand that:
* passing newline characters through the path is some kind of edge case and not what people usually do;
* some weird solutions are required to support this edge case and the profit is disputable;
* it could probably be better to restrict this data in the application so that it only contains characters from some allowed subset of Unicode characters, but the API is already established and it is out of the scope of this issue.

Anyway, I think that you should know about the issue. Does it worth to fix it? What do you think?

### To reproduce

```python

def test_starlette_route_with_new_line_characters() -> None:
    from starlette.routing import Route
    from starlette.middleware.wsgi import WSGIMiddleware

    # the second argument is not important, it can be anything
    route = Route("/{path:path}", WSGIMiddleware(lambda x: None))

    # everything works as expected without newline characters
    match, child_scope = route.matches({"type": "http", "path": "/qwe"})
    assert match is Match.FULL
    assert child_scope["path_params"]["path"] == "qwe"

    # the single new line character at the end is cut off
    match, child_scope = route.matches({"type": "http", "path": "/qwe\n"})
    assert match is Match.FULL
    assert child_scope["path_params"]["path"] == "qwe"

    # route doesn't match if you put two newline characters
    match, child_scope = route.matches({"type": "http", "path": "/qwe\n\n"})
    assert match is Match.NONE

    # route doesn't match if you put the newline character in the middle
    match, child_scope = route.matches({"type": "http", "path": "/qwe\nrty"})
    assert match is Match.NONE
```

### Expected behavior

Paths can contain any character (at least, in URL encoded form) and correctly match the route.

### Actual behavior

* In some cases, Starlette is not able to match the route with the URL containing newline characters;
* In other cases, Starlette corrupts the data by cutting newline characters off.

### Environment

- OS: macOS
- Python version: 3.8.3
- Starlette version: 0.13.8 (get it with `$ pip show starlette`)

**Comment by lovelydinosaur** [comment]
Yeah I think let's just roll with not supporting this.
If anyone does need it, then a custom convertor would be the way to go.

**Comment by Pentusha** [comment]
Please check my fix for this issue.
@tomchristie it's not about the converter, but about incorrect work with regular expressions.
Check this case please. #1290

**Comment by Kludex** [comment]
Closed by #1437 .

It will be available on Starlette 0.19.0.
